rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/NodeSourcePluginManager.java

```diff
@@ -0,0 +1,171 @@
+ * as published by the Free Software Foundation: version 3 of
+ * the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * If needed, contact us to obtain a release under GPL Version 2 or 3
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive.resourcemanager.core;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import org.ow2.proactive.resourcemanager.nodesource.NodeSourceDescriptor;
+import org.ow2.proactive.resourcemanager.nodesource.common.ConfigurableField;
+import org.ow2.proactive.resourcemanager.nodesource.common.PluginDescriptor;
+import org.ow2.proactive.resourcemanager.nodesource.infrastructure.InfrastructureManager;
+import org.ow2.proactive.resourcemanager.nodesource.policy.NodeSourcePolicy;
+
+
+public class NodeSourcePluginManager {
+
+    public Collection<ConfigurableField> getPolicyConfigurableFields(String nodeSourceName,
+            NodeSourceDescriptor descriptor) {
+
+        Class<NodeSourcePolicy> nodeSourcePolicyClass = this.getNodeSourcePolicyClassOrFail(nodeSourceName, descriptor);
+        PluginDescriptor policyPluginDescriptor = new PluginDescriptor(nodeSourcePolicyClass, new HashMap<>());
+
+        return policyPluginDescriptor.getConfigurableFields();
+    }
+
+    public Collection<ConfigurableField> getInfrastructureConfigurableFields(String nodeSourceName,
+            NodeSourceDescriptor descriptor) {
+
+        Class<InfrastructureManager> infrastructureManagerClass = this.getInfrastructureManagerClassOrFail(nodeSourceName,
+                                                                                                           descriptor);
+        PluginDescriptor infrastructurePluginDescriptor = new PluginDescriptor(infrastructureManagerClass,
+                                                                               new HashMap<>());
+
+        return infrastructurePluginDescriptor.getConfigurableFields();
+    }
+
+    public List<Serializable> getParametersWithDynamicUpdated(Object[] newParameters, List<Serializable> oldParameters,
+            Collection<ConfigurableField> configurableFields) {
+
+        List<Serializable> parametersWithDynamicUpdated = new LinkedList<>();
+        parametersWithDynamicUpdated.addAll(oldParameters);
+
+        int oldValueIndex;
+        String newValue;
+        String oldValue;
+
+        for (int newValueIndex = 0; newValueIndex < newParameters.length; newValueIndex++) {
```

Comment:
You can create forEachWithIndex which takes 3 collections.

Reply:
this loop looks a bit complex. Can we refactor it a bit ?

Reply:
Yes, this is done. I created a forEachWithIndex, in Lambda class, but only with one collection, as it is more likely to be reused this way.